### ISSUE TITLE
[RMP] Support Tree Ranking Models (like XGBoost) in Merlin Models and Systems

### ISSUE THREAD

**Issue by viswa-nvidia** [NVIDIA-Merlin/Merlin#105]
## Problem

Gradient-boosted decision trees (GBDTs) are commonly used in the industry as part of the scoring phase of recommender systems. Supporting serving of these models and integrating with the Merlin ecosystem will help facilitate usage of these models in these systems.

The Triton Inference Server has a [backend called FIL (Forest Inference Library)](https://github.com/triton-inference-server/fil_backend) to facilitate GPU accelerated serving of these models.

> Random forests (RF) and gradient-boosted decision trees (GBDTs) have become workhorse models of applied machine learning. XGBoost and LightGBM, popular packages implementing GBDT models, consistently rank among the most commonly used tools by data scientists on the Kaggle platform. We see similar interest in forest-based models in industry, where they are applied to problems ranging from inventory forecasting, to ad ranking, to medical diagnostics.
>
> [RAPIDS Forest Inference Library: Prediction at 100 million rows per second](https://medium.com/rapids-ai/rapids-forest-inference-library-prediction-at-100-million-rows-per-second-19558890bc35)

## Goals

- Enable the use of Tree based models (e.g. GBDTs, Random Forests) in a Merlin Systems ensemble.
- Support the training of XGBoost models from a Merlin Dataset.

## Constraints

- Use the Triton [FIL backend](https://github.com/triton-inference-server/fil_backend) in Merlin systems

### Merlin-models (Data Scientist)

- [x] https://github.com/NVIDIA-Merlin/models/issues/132
  - First version of API
     - https://github.com/NVIDIA-Merlin/models/pull/433 
     - https://github.com/NVIDIA-Merlin/models/pull/466
  - Example
     - https://github.com/NVIDIA-Merlin/models/pull/522

### NVTabular (Data Scientist)

- [NA] Operators for batch prediction with these models 
- Note: Batch prediction is not in scope for this development

### Merlin-systems (Product Engineer)

- [X] FIL integration in ensemble API.
  Supports the following models: XGBoost, LightGBM, Scikit-Learn (Random Forest), cuML (Random Forest)
  - https://github.com/NVIDIA-Merlin/systems/pull/118
  - https://github.com/NVIDIA-Merlin/systems/pull/110

### Examples and Docs (Everyone)
- [x] Example of training  tree models ( End to End example )
- [ ] NVIDIA-Merlin/models#575

Aha! Link: https://nvaiinfa.aha.io/features/MERLIN-828

**Comment by viswa-nvidia** [comment]
@benfred , could you map your XGboost tickets here ?

**Comment by karlhigley** [comment]
@marcromeyn Re-assigned this to you for the time being since it sounds like you'll be working with our new hire on both the Models and Systems sides. Once he's in our GH org, might make him the lead on this though?

**Comment by oliverholworthy** [comment]
An update on this issue.

_Merlin Models_

We have a first version of the XGBoost API in Merlin Models from release 22.06. Once we have an example that part can be considered complete, for the purposes of this issue. With ongoing support for feature improvements via other issues.

_Merlin Systems_

Operators have been added to Merlin Systems to enable XGBoost, LightGBM, Scikit-Learn (Random Forest), cuML (Random Forest) models being added as part of a serving ensemble in Triton. There are a couple of small issues being addressed (operator outputs) to make this usable from the next release 22.07.

_NVTabular_

The comment "NVTabular - Operators for batch prediction with these models" in the description I think may be out of scope of this issue. At least, it's unclear to me what the relationship with NVTabular is in this context. In terms of batch prediction for the XGBoost integration in Merlin Models. This is supported as part of the predict method, since we're using the Dask version of XGBoost, which can be used in a distributed setting. 

In terms of a common pattern for running evaluation / which will involve batch/distributed prediction. The ongoing work in this area may be covered by https://github.com/NVIDIA-Merlin/Merlin/issues/407 or https://github.com/NVIDIA-Merlin/Merlin/issues/405

**Comment by viswa-nvidia** [comment]
@radekosmulski , could you please add the ticket tracking the examples here in the top

**Comment by viswa-nvidia** [comment]
Closing this ticket. Example is pending and is planned for 22.08
